app/src/main/java/com/mikhailovskii/weatherandroid/data/firebase/FirebaseModel.kt

```diff
@@ -0,0 +1,188 @@
+        }.addOnFailureListener {
+            callback?.onLoginFailed()
+        }
+
+    }
+
+    fun logInWithFacebook(user: User, callback: FirebaseLoginCallback?) {
+        db.collection(USERS_COLLECTION).get()
+            .addOnSuccessListener { databaseResult ->
+                var isUserPresents = false
+
+                loop@ for (document in databaseResult) {
+                    val databaseUser = document.toObject(User::class.java)
+
+                    if (databaseUser.login == user.login
+                        && databaseUser.facebookKey == user.facebookKey
+                        && databaseUser.icon == user.icon
+                    ) {
+                        isUserPresents = true
+                        break@loop
+                    }
+                }
+
+                if (!isUserPresents) {
+                    db.collection(USERS_COLLECTION).document().set(user)
+                        .addOnSuccessListener {
+                            Timber.d(("Facebook info saved"))
+                        }.addOnFailureListener { e ->
+                            Timber.e("Error facebook save: $e")
+                        }
+                }
+
+                callback?.onLoggedInWithFilledInfo()
+            }.addOnFailureListener {
+                callback?.onLoginFailed()
+            }
+    }
+
+    fun logInWithGoogle(user: User, callback: FirebaseLoginCallback?) {
+        db.collection(USERS_COLLECTION).get().addOnSuccessListener { databaseResult ->
+            var isUserPresents = false
+
+            loop@ for (document in databaseResult) {
+                val databaseUser = document.toObject(User::class.java)
+
+                if (databaseUser.login == user.login
+                    && databaseUser.googleKey == user.googleKey
+                    && databaseUser.icon == user.icon
+                ) {
+                    isUserPresents = true
+                    break@loop
+                }
+            }
+
+            if (!isUserPresents) {
+                db.collection(USERS_COLLECTION).document().set(user)
+                    .addOnSuccessListener {
+                        Timber.d(("Facebook info saved"))
+                    }.addOnFailureListener { e ->
+                        Timber.e("Error facebook save: $e")
+                    }
+            }
+
+            callback?.onLoggedInWithFilledInfo()
+        }.addOnFailureListener {
+            callback?.onLoginFailed()
+        }
+    }
+
+    fun getStickerPackByName(name: String, callback: FirebaseDataCallback?) {
+        db.collection(STICKERS_COLLECTION)
+            .whereEqualTo(TITLE_FIELD, name)
+            .get()
+            .addOnSuccessListener { result ->
+                if (!result.isEmpty) {
+                    for (document in result) {
+                        val stickerPack = document.toObject(StickerPack::class.java)
+                        callback?.onFirebaseDataLoaded(stickerPack)
```

Comment:
@SergeiMikhailovskii is it okay that you have `n` events `onFirebaseDataLoaded()` (one per each document), but only one event `onFirebaseDataFailed()` (one per request)?

Reply:
@anton-novikau document always has only one element, so there is always only one successful call or one fail call. I found strange such construction, but I saw it in a big amount of tutorials and that's why I used it in this way

Reply:
then you should handle it in a different way. Instead of `foreach` you should obtain the very first element (if it doesn't support `get(int)`, you may use `result.iterator.next()`) or you can continue using `foreach`, but you should break the cycle after the very first iteration.

Reply:
@SergeiMikhailovskii any update here?

Reply:
@anton-novikau sorry, didn't notice this comment. Fixed this in this way: `result.iterator.next()` and merged to master